File: leetcode207/src/Solution.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        ArrayList<Integer>[] g = new ArrayList[numCourses];
        Arrays.setAll(g,e->new ArrayList<>());
        int[] inDeg = new int[numCourses];
        for (int[] pre : prerequisites) {
            g[pre[1]].add(pre[0]);
            inDeg[pre[0]]++;//将需要先修课的课程的入度+1
        }
        var q = new ArrayDeque<Integer>();
        for (int i = 0; i < numCourses; i++) {
            if (inDeg[i] == 0) q.addLast(i);//添加所有入度为0的点，即不需要先修课的点
        }
        while (!q.isEmpty()) {
            int x = q.pollFirst();
            //修完课程x，此时所有需要x作为先修课的课程y的入度-1
            for (Integer y : g[x]) {
                inDeg[y]--;
                if (inDeg[y] == 0) q.addLast(y);
            }
        }
        for (int i = 0; i < numCourses; i++) {
            if (inDeg[i] != 0) return false;
        }
        return true;
    }
}

File: leetcode207/Solution.py
from collections import deque
from typing import List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        g = [[] for _ in range(numCourses)]
        inDeg = [0 for _ in range(numCourses)]
        for pre in prerequisites:
            g[pre[1]].append(pre[0])
            inDeg[pre[0]] += 1  # 将需要先修课的课程的入度+1
        q = deque()
        for i in range(numCourses):
            if inDeg[i] == 0:  # 添加所有入度为0的点，即不需要先修课的点
                q.append(i)
        while q:
            x = q.popleft()
            # 修完课程x，此时所有需要x作为先修课的课程y的入度-1
            for y in g[x]:
                inDeg[y] -= 1
                if inDeg[y] == 0:
                    q.append(y)

        for i in range(numCourses):
            if inDeg[i] != 0:
                return False
        return True
